src/protocols/screencopy.rs

```diff
@@ -0,0 +1,279 @@
+    zwlr_screencopy_frame_v1, zwlr_screencopy_manager_v1,
+};
+use smithay::reexports::wayland_server::protocol::wl_buffer::WlBuffer;
+use smithay::reexports::wayland_server::protocol::wl_shm;
+use smithay::reexports::wayland_server::{
+    Client, DataInit, Dispatch, DisplayHandle, GlobalDispatch, New, Resource,
+};
+use smithay::utils::{Physical, Rectangle};
+
+const VERSION: u32 = 3;
+
+pub struct ScreencopyManagerState;
+
+impl ScreencopyManagerState {
+    pub fn new<D>(display: &DisplayHandle) -> Self
+    where
+        D: GlobalDispatch<ZwlrScreencopyManagerV1, ()>,
+        D: Dispatch<ZwlrScreencopyManagerV1, ()>,
+        D: Dispatch<ZwlrScreencopyFrameV1, ScreencopyFrameState>,
+        D: ScreencopyHandler,
+        D: 'static,
+    {
+        display.create_global::<D, ZwlrScreencopyManagerV1, _>(VERSION, ());
+
+        Self
+    }
+}
+
+impl<D> GlobalDispatch<ZwlrScreencopyManagerV1, (), D> for ScreencopyManagerState
+where
+    D: GlobalDispatch<ZwlrScreencopyManagerV1, ()>,
+    D: Dispatch<ZwlrScreencopyManagerV1, ()>,
+    D: Dispatch<ZwlrScreencopyFrameV1, ScreencopyFrameState>,
+    D: ScreencopyHandler,
+    D: 'static,
+{
+    fn bind(
+        _state: &mut D,
+        _display: &DisplayHandle,
+        _client: &Client,
+        manager: New<ZwlrScreencopyManagerV1>,
+        _manager_state: &(),
+        data_init: &mut DataInit<'_, D>,
+    ) {
+        data_init.init(manager, ());
+    }
+}
+
+impl<D> Dispatch<ZwlrScreencopyManagerV1, (), D> for ScreencopyManagerState
+where
+    D: GlobalDispatch<ZwlrScreencopyManagerV1, ()>,
+    D: Dispatch<ZwlrScreencopyManagerV1, ()>,
+    D: Dispatch<ZwlrScreencopyFrameV1, ScreencopyFrameState>,
+    D: ScreencopyHandler,
+    D: 'static,
+{
+    fn request(
+        _state: &mut D,
+        _client: &Client,
+        manager: &ZwlrScreencopyManagerV1,
+        request: zwlr_screencopy_manager_v1::Request,
+        _data: &(),
+        _display: &DisplayHandle,
+        data_init: &mut DataInit<'_, D>,
+    ) {
+        let (frame, overlay_cursor, rect, output) = match request {
+            zwlr_screencopy_manager_v1::Request::CaptureOutput {
+                frame,
+                overlay_cursor,
+                output,
+            } => {
+                let output = Output::from_resource(&output).unwrap();
+                let rect =
+                    Rectangle::from_loc_and_size((0, 0), output.current_mode().unwrap().size);
+                (frame, overlay_cursor, rect, output)
+            }
+            zwlr_screencopy_manager_v1::Request::CaptureOutputRegion {
+                frame,
+                overlay_cursor,
+                x,
+                y,
+                width,
+                height,
+                output,
+            } => {
+                let output = Output::from_resource(&output).unwrap();
+                let rect = Rectangle::from_loc_and_size((x, y), (width, height));
```

Comment:
Is anything gonna break if the client sends invalid coordinates like negative width / height?

Reply:
Hm, i hadn't even thought about that. The other implementations i looked at don't seem to validate this, but it might be an issue? I'm unsure

Reply:
in debug mode: niri crashes. smithay has debug asserts to enure the size should never be negative.
in release mode: i believe the cast to `u32` results in niri telling the client that the buffer should allocate  buffer with approximately `u32::MAX - width` width (or vice versa for height, stride). niri does not crash, but the client i was testing this with (modified wayshot to allocate a negative width) does, and niri logs an error in client communication. if the client somehow detects this and allocates a better buffer, niri will reject it for being wrong (comparing to the requested region).. if the client somehow allocates the requested buffer without dying, then both the height *and* width must be negative for their product to ever be positive.

i'm gonna normalize the requested region such that negative width/height means the region extents up and left from the given point.

additionally, what about negative/offscreen coordinates, such that the resulting region is not within screen bounds? well, for these, they are simply cropped to the screen bounds, and niri will request a buffer of different specs than the given width/height that the client asked for.

Reply:
Doesn't it make more sense to just fail the request if width or height are <= 0?